primitives/npos-elections/src/reduce.rs

```diff
@@ -417,8 +416,8 @@ fn reduce_all<A: IdentifierT>(assignments: &mut Vec<StakedAssignment<A>>) -> u32
 				// find minimum of cycle.
 				let mut min_value: ExtendedBalance = Bounded::max_value();
 				// The voter and the target pair that create the min edge.
-				let mut min_target: A = Default::default();
-				let mut min_voter: A = Default::default();
+				let mut min_target: Option<A> = None;
```

Comment:
About this: after looking at the code, this bit seems like the only controversial bit. My conclusion is that this diff is not changing the logic of this snippet at all. The `Default::default()`, and `min_index = 0` were _clearly meant_ to be overwritten in the `for loop` that follows. In fact, the outcome is (pretty bad) UB if it doesn't. The changes of this PR are not altering any of that. The assumption still must hold that these values get written to. Making them an `Option<_>` doesn't harm that.

I wish I had the time now, or in the near future, to re-write the majority of this `reduce` algorithm. Admittedly, it is the last historical Spaghetti code left in `sp-npos-election` that I myself even have a very hard time reading. It was mostly written during a crunch and never been re-visited ever since. While I can't take the time now to improve this code fundamentally, I did propose some reasonable changes in [this branch](https://github.com/paritytech/substrate/compare/gav-no-default-accountid...gav-no-default-accountid-staking?expand=1) to make it more compatible with `Option<_>`, most notably making the assumption explicit that these `Option` will all become `Some` and stop the algorithm otherwise.

Lastly, luckily this code has a fuzzer that has proven quite helpful in the past. I'll re-run it again on my aforementioned branch and ring some bells if anything is broken. 

Feel free to either apply the commits of https://github.com/paritytech/substrate/compare/gav-no-default-accountid...gav-no-default-accountid-staking?expand=1 either directly here, or I can open a PR on top of this. All changes in it are pretty clear in my opinion.